arwen/cryptoapi/cryptoei.go

```diff
@@ -287,3 +372,607 @@ func v1_4_verifySecp256k1(
+		return 1
+	}
+	oneByteScalarGasCost := metering.GasSchedule().CryptoAPICost.ScalarMultECC * uint64(curveMultiplier) / 100
+	gasToUse := oneByteScalarGasCost + uint64(length)*oneByteScalarGasCost
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	ec, err1 := managedType.GetEllipticCurve(ecHandle)
+	if err1 != nil {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	xResult, yResult, err1 := managedType.GetTwoBigInt(xResultHandle, yResultHandle)
+	x, y, err2 := managedType.GetTwoBigInt(pointXHandle, pointYHandle)
+	if err1 != nil || err2 != nil {
+		arwen.WithFault(arwen.ErrNoBigIntUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	if !ec.IsOnCurve(x, y) {
+		arwen.WithFault(arwen.ErrPointNotOnCurve, context, runtime.CryptoAPIErrorShouldFailExecution())
+	}
+
+	managedType.ConsumeGasForBigIntCopy(xResult, yResult, ec.P, ec.N, ec.B, ec.Gx, ec.Gy, x, y)
+	xResultSM, yResultSM := ec.ScalarMult(x, y, data)
+	xResult.Set(xResultSM)
+	yResult.Set(yResultSM)
+
+	return 0
+}
+
+//export v1_4_marshalEC
+func v1_4_marshalEC(
+	context unsafe.Pointer,
+	xPairHandle int32,
+	yPairHandle int32,
+	ecHandle int32,
+	resultOffset int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	curveMultiplier := managedType.Get100xCurveGasCostMultiplier(ecHandle)
+	if curveMultiplier < 0 {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	gasToUse := metering.GasSchedule().CryptoAPICost.MarshalECC * uint64(curveMultiplier) / 100
+	metering.UseGas(gasToUse)
+
+	ec, err := managedType.GetEllipticCurve(ecHandle)
+	if err != nil {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	x, y, err := managedType.GetTwoBigInt(xPairHandle, yPairHandle)
+	if err != nil {
+		arwen.WithFault(arwen.ErrNoBigIntUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	if !ec.IsOnCurve(x, y) {
+		arwen.WithFault(arwen.ErrPointNotOnCurve, context, runtime.CryptoAPIErrorShouldFailExecution())
+	}
+	if x.BitLen() > int(ec.BitSize) || y.BitLen() > int(ec.BitSize) {
+		arwen.WithFault(arwen.ErrLengthOfBufferNotCorrect, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	managedType.ConsumeGasForBigIntCopy(ec.P, ec.N, ec.B, ec.Gx, ec.Gy, x, y)
+	result := elliptic.Marshal(ec, x, y)
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return int32(len(result))
+	}
+	return 0
```

Comment:
These `return`s should be the other way around. Return `len(result)` on success, return 1 on failure.